### PR TITLE
BrushIntervalSelector

### PR DESCRIPTION
**Pull request by martinRenou**
Fix https://github.com/bqplot/bqplot/issues/1180

See https://github.com/d3/d3-brush/issues/43

### CODE REVIEW

**Comment by maartenbreddels** [comment]
See also https://github.com/bqplot/bqplot/pull/1097

**Comment by martinRenou** [comment]
@kaiayoung is it the issue you were mentioning?

**Comment by maartenbreddels** [comment]
This looks similar to what I've done in #1097, although that also contains another change extend->move.

Could you dig into why that was needed?

**Comment by martinRenou** [comment]
I am actually not sure this is needed, because the move is done later on in `syncModelToBrush`

File: js/src/BrushSelector.ts [hunk 1]
```diff
@@ -317,7 +317,6 @@ export class BrushSelector extends BrushMixinXYSelector {
 
         this.brush.extent([[0, 0], [this.width, this.height]]);
         this.syncModelToBrush();
-        this.brushsel = this.d3el.call(this.brush);
```
Comment (martinRenou) [inline js/src/BrushSelector.ts:320]: As discussed in https://github.com/bqplot/bqplot/pull/1097

**Comment by martinRenou** [comment]
@maartenbreddels do you think it's ready for merging?

**Comment by ibdafna** [comment]
@martinRenou The plot renders in the second tab, but it looks like the bouncing is still there unfortunately. 

bar_chart_selector_bouncing

```python
from bqplot import (OrdinalScale, LinearScale, Bars, Figure, Axis, ColorScale, CATEGORY10)
from bqplot.interacts import BrushIntervalSelector
import ipywidgets as widgets
import bqplot as bqp
import pandas as pd
import numpy as np
np.random.seed(0)
plt_x = np.arange(10)
plt_y = np.random.randint(10, size=10)

def print_val(change):
    print(change['new'])
        
x_sc = OrdinalScale()
y_sc = LinearScale()

plot = Bars(x=plt_x, y=plt_y, scales={'x':x_sc, 'y':y_sc}, colors=['#ff5500'],
            labels=['Date', 'Reads'])

xax = Axis(scale=x_sc, grid_lines='none', num_ticks=10)
yax = Axis(scale=y_sc, tick_format=',d', grid_lines='none', orientation='vertical', min=0)

brushsel_date = BrushIntervalSelector(scale=x_sc, marks=[plot], color='skyblue')

plot.observe(print_val, names=['selected'])

fig = Figure(marks=[plot], axes=[xax, yax], 
                              padding_x=0.025, padding_y=0.025,
                             interaction=brushsel_date)


tab = widgets.Tab()
tab.children = [fig, fig]
# widgets.VBox([fig, fig])
tab
```

**Comment by martinRenou** [comment]
This looks like another issue doesn't it? This PR fixes the resize of the BrushSelector https://github.com/bqplot/bqplot/issues/1180

I am looking at the bouncing between views issue, seems very related to https://github.com/bqplot/bqplot/pull/1098, there is so much code duplication in this library I would not be surprised we simply need to apply the same fix at another class.

**Comment by ibdafna** [comment]
They're closely related, I think. If you look at #1180, the title says it is only visible initially but disappears when used in a tab or accordion, which arguably includes what we're seeing above. Originally the issue was with the selection only showing once as you say due to invalid figure size. But the current result is probably worse as you don't even end up with a single selection due to the bouncing. Happy to open a separate PR if needed!

**Comment by martinRenou** [comment]
Ok, now I see why they are both related. Thanks.

I'm investigating a good way to fix this.

**Comment by martinRenou** [comment]
I feel like the bouncing is due to the `this.brushsel = this.d3el.call(this.brush);` line which triggers `brush_start` and `brush_end`. This line is called when the selection changes for updating the brush position, and the fact that it triggers those events it might update the selection as well.

**Comment by martinRenou** [comment]
There are multiple issues with the BrushIntervalSelector (some of them appear on the BrushSelector as well, but let's deal with them one at a time):

- [x] Updating the selector when a change appeared in another view/in the model needs to be done my calling `this.brush.move` and `this.d3el.call(this.brush)`. This results in triggering `brush_start` and `brush_move`, which itself will update the model and redraw the brush -> lots of bouncing. *This PR fixes this issue by discarding the events during the selector redraw*
- [x] The `selected` attribute was not using TypedArrays, resulting in bouncing between the Python model and the JavaScript one.
- [ ] The first `relayout` in a tab does not seem to work, which is very odd because `relayout` appears to work properly on window resizing.
- [ ] BrushIntervalSelector with an ordinal scale does not work. It's either coming from issues in the OrdinalScale itself, either coming from how the BrushIntervalSelector uses the scale, or it's coming from the `selected` attribute which does not have the right TypedArray

In this PR I use a `Float32Array` for the `selected` attribute, this seems like an odd choice to me, as we might not know the type of the buffer in advance, it depends on the scale or the data I guess.

**Comment by martinRenou** [comment]
We'll need to apply similar fixes to the BrushSelector, and maybe the MultiSelector

**Comment by ibdafna** [comment]
LGTM! Thank you

**Pull request merged**